lib/readline.js

```diff
@@ -580,27 +591,57 @@ Interface.prototype._wordLeft = function() {
 Interface.prototype._wordRight = function() {
   if (this.cursor < this.line.length) {
     var trailing = this.line.slice(this.cursor);
-    var match = trailing.match(/^(?:\s+|\W+|\w+)\s*/);
+    var match = trailing.match(/^(?:\s+|[^\w\s]+|\w+)\s*/);
     this._moveCursor(match[0].length);
   }
 };
 
+function codePointLeftOf(i, str) {
+  // obtain the code point to the left
+  let previousChar;
+  let len = 0;
+  for (const char of str) {
+    len += char.length;
+    previousChar = char;
+    if (len === i) {
+      return previousChar;
+    }
+  }
+}
+
+function codePointRightOf(i, str) {
+  // obtain the code point to the right
+  let len = 0;
+  for (const char of str) {
+    len += char.length;
+    if (len > i) {
+      // UTF-16 units comprising the character
+      return char;
+    }
+  }
+}
```

Comment:
This could be rewritten to `O(1)`:

```js
function charLengthAt(str, i) {
  if (str.length <= i)
    return 0;
  return str.codePointAt(i) > 2 ** 16 ? 2 : 1;
}
```